File: golang/pkg/agentbay/lifecycle_policy.go
package agentbay

import (
	"fmt"
)

// LifecyclePolicy configures session lifecycle: idle release, max runtime, and optional manual release.
type LifecyclePolicy struct {
	// IdleReleaseTimeout is the idle release timeout in minutes.
	IdleReleaseTimeout int32
	// MaxRuntime is the maximum session runtime in minutes.
	MaxRuntime int32
	// ManualRelease when true means the session is released only by explicit user action.
	ManualRelease bool
}

// NewLifecyclePolicy returns defaults: 5 minutes idle release, 30 minutes max runtime, manual release off.
func NewLifecyclePolicy() *LifecyclePolicy {
	return &LifecyclePolicy{
		IdleReleaseTimeout: 5,
		MaxRuntime:         30,
		ManualRelease:      false,
	}
}

// NewLifecyclePolicyManualRelease returns manual release mode (idle and max runtime are unused).
func NewLifecyclePolicyManualRelease() *LifecyclePolicy {
	return &LifecyclePolicy{
		IdleReleaseTimeout: 0,
		MaxRuntime:         0,
		ManualRelease:      true,
	}
}

// NewLifecyclePolicyWithValues validates and returns a LifecyclePolicy.
// When manualRelease is true, idleReleaseTimeout and maxRuntime must be zero.
// When manualRelease is false, idleReleaseTimeout and maxRuntime must be positive.
func NewLifecyclePolicyWithValues(idleReleaseTimeout, maxRuntime int32, manualRelease bool) (*LifecyclePolicy, error) {
	if manualRelease {
		if idleReleaseTimeout != 0 || maxRuntime != 0 {
			return nil, fmt.Errorf("manual release mode cannot set idle release timeout or max runtime")
		}
		return &LifecyclePolicy{
			IdleReleaseTimeout: 0,
			MaxRuntime:         0,
			ManualRelease:      true,
		}, nil
	}
	if idleReleaseTimeout <= 0 || maxRuntime <= 0 {
		return nil, fmt.Errorf("idle release timeout and max runtime must be positive integers")
	}
	return &LifecyclePolicy{
		IdleReleaseTimeout: idleReleaseTimeout,
		MaxRuntime:         maxRuntime,
		ManualRelease:      false,
	}, nil
}

File: java/agentbay/src/main/java/com/aliyun/agentbay/session/LifecyclePolicy.java
package com.aliyun.agentbay.session;

/**
 * Lifecycle policy for session management.
 *
 * Controls how and when a session is automatically released.
 * When used, SDK takes full control of lifecycle - console defaults are overridden.
 * All time values are in MINUTES.
 */
public class LifecyclePolicy {

    private int idleReleaseTimeout;
    private int maxRuntime;
    private boolean manualRelease;

    /**
     * Default constructor: idle=5min, max=30min, manual=false
     */
    public LifecyclePolicy() {
        this.idleReleaseTimeout = 5;
        this.maxRuntime = 30;
        this.manualRelease = false;
    }

    /**
     * Custom timeouts constructor.
     *
     * @param idleReleaseTimeout idle timeout in minutes (must be positive)
     * @param maxRuntime maximum runtime in minutes (must be positive)
     */
    public LifecyclePolicy(int idleReleaseTimeout, int maxRuntime) {
        if (idleReleaseTimeout <= 0) {
            throw new IllegalArgumentException("idleReleaseTimeout must be a positive integer (minutes)");
        }
        if (maxRuntime <= 0) {
            throw new IllegalArgumentException("maxRuntime must be a positive integer (minutes)");
        }
        this.idleReleaseTimeout = idleReleaseTimeout;
        this.maxRuntime = maxRuntime;
        this.manualRelease = false;
    }

    /**
     * Full constructor with manual release validation.
     *
     * @param idleReleaseTimeout idle timeout in minutes
     * @param maxRuntime maximum runtime in minutes
     * @param manualRelease when true, idle and max cannot be non-zero defaults
     */
    public LifecyclePolicy(int idleReleaseTimeout, int maxRuntime, boolean manualRelease) {
        if (manualRelease) {
            if (idleReleaseTimeout != 0 || maxRuntime != 0) {
                throw new IllegalArgumentException(
                    "idleReleaseTimeout and maxRuntime cannot be set when manualRelease=true");
            }
            this.idleReleaseTimeout = 0;
            this.maxRuntime = 0;
            this.manualRelease = true;
            return;
        }
        if (idleReleaseTimeout <= 0) {
            throw new IllegalArgumentException("idleReleaseTimeout must be a positive integer (minutes)");
        }
        if (maxRuntime <= 0) {
            throw new IllegalArgumentException("maxRuntime must be a positive integer (minutes)");
        }
        this.idleReleaseTimeout = idleReleaseTimeout;
        this.maxRuntime = maxRuntime;
        this.manualRelease = false;
    }

    /**
     * Factory method for manual release mode.
     *
     * @return policy with manual release only (no automatic idle/max limits)
     */
    public static LifecyclePolicy manualRelease() {
        LifecyclePolicy lp = new LifecyclePolicy();
        lp.idleReleaseTimeout = 0;
        lp.maxRuntime = 0;
        lp.manualRelease = true;
        return lp;
    }

    public int getIdleReleaseTimeout() {
        return idleReleaseTimeout;
    }

    public int getMaxRuntime() {
        return maxRuntime;
    }

    public boolean isManualRelease() {
        return manualRelease;
    }
}
